server/src/gnome/prompter.rs

```diff
@@ -0,0 +1,205 @@
+            PromptRole::Lock => {
+                if properties.is_empty() {
+                    // First PromptReady call
+                    let mut properties: HashMap<&str, zvariant::OwnedValue> = HashMap::new();
+                    let secret_exchange = SecretExchange::new().map_err(|err| {
+                        ServiceError::ZBus(zbus::Error::FDO(Box::new(zbus::fdo::Error::Failed(
+                            format!("Failed to generate SecretExchange {err}."),
+                        ))))
+                    })?;
+                    let exchange = secret_exchange.begin();
+
+                    properties.insert("continue-label", Value::new("Lock").try_to_owned().unwrap());
+                    properties.insert(
+                        "description",
+                        Value::new("Confirm locking \"login\", Keyring.")
+                            .try_to_owned()
+                            .unwrap(),
+                    );
+                    // TODO: set the object name dynamically in description
+                    properties.insert(
+                        "message",
+                        Value::new("Lock Keyring").try_to_owned().unwrap(),
+                    );
+                    properties.insert("caller-window", Value::new("").try_to_owned().unwrap());
+                    properties.insert("cancel-label", Value::new("Cancel").try_to_owned().unwrap());
+
+                    let connection = connection.clone();
+                    let path = self.path.clone();
+
+                    tokio::spawn(PrompterCallback::perform_prompt(
+                        connection, path, "confirm", properties, exchange,
+                    ));
+                } else if reply == "no" {
+                    // Second PromptReady call and the prompt is dismissed
+                    tracing::debug!("Prompt is being dismissed.");
+
+                    tokio::spawn(PrompterCallback::stop_prompting(
+                        connection.clone(),
+                        self.path.clone(),
+                    ));
+
+                    let signal_emitter = self.service.signal_emitter(prompt.path().clone())?;
+                    let result = Value::new::<Vec<OwnedObjectPath>>(vec![])
+                        .try_to_owned()
+                        .unwrap();
+
+                    tokio::spawn(PrompterCallback::prompt_completed(
+                        signal_emitter,
+                        true,
+                        result,
+                    ));
+                } else {
+                    // Second PromptReady call with the final exchange
+                    let service = self.service.clone();
+                    let objects = prompt.objects().clone();
+                    let result = Value::new(&objects).try_to_owned().unwrap();
+
+                    tokio::spawn(async move {
+                        let _ = service.set_locked(true, &objects, true).await;
+                    });
```

Comment:
Instead of passing the objects, this should pass a future directly that you can spawn here. Keeping the logic that is supposed to be in Service::set_locked, stay there.

Reply:
https://github.com/bilelmoussaoui/oo7/pull/173#discussion_r1966813203